cli/lightning-cli.c

```diff
@@ -95,20 +96,88 @@ static size_t human_readable(const char *buffer, const jsmntok_t *t, char term)
 }
 
 static void human_help(const char *buffer, const jsmntok_t *result, bool has_command) {
-	int i;
-	const jsmntok_t * help_array = result + 2;
-	/* the first command object */
-	const jsmntok_t * curr = help_array + 1;
-	/* iterate through all commands, printing the name and description */
-	for (i = 0; i < help_array->size; i++) {
-		curr += 2;
-		printf("%.*s\n", curr->end - curr->start, buffer + curr->start);
-		curr += 2;
-		printf("    %.*s\n\n", curr->end - curr->start, buffer + curr->start);
-		curr += 2;
-		/* advance to next command */
-		curr++;
+	unsigned int i, j;
+	/* `curr` is used as a temporary token and `command_category` in case of error */
+	const jsmntok_t *curr, *command_category;
+	/* Contains all commands objects, which have the following structure :
+	 * {
+	 *     "command": "The command name",
+	 *     "category": "The command category",
+	 *     "description": "The command's description",
+	 *     "verbose": "The command's detailed description"
+	 * }
+	 */
+	const jsmntok_t * help_array = json_get_member(buffer, result, "help");
+	/* We will sort commands by populating an array of json token pointers by category.
+	 * We create an array which contains them all to avoid redontant iteration through
+	 * each category array and to simplify memory management (tal_free(top_array) frees
+	 * all contained arrays
+	 */
+	const jsmntok_t ***commands = tal_arr(tmpctx, const jsmntok_t**, 7);
+	for (i = 0; i < tal_count(commands); ++i) {
+		commands[i] = tal_arr(tmpctx, const jsmntok_t*, 0);
 	}
+	/* A convenient mapping to compare categories in command objects (string) with our
+	 * categories enum, and to display category names when iterating through the arrays
+	 */
+	const char **category_names = tal_arr(tmpctx, const char*, 7);
```

Comment:
These can likely be statically initialized and do not need to be allocated on the heap. It'd also allow us to use `ARRAY_SIZE` instead of the constant `7` in the `tal_arr` initialization above.